File: ios-extensions/NotificationService.m
//  NotificationService.m
//  ShopUnite Notification Service Extension

#import "NotificationService.h"

@interface NotificationService ()

@property (nonatomic, strong) void (^contentHandler)(UNNotificationContent *contentToDeliver);
@property (nonatomic, strong) UNNotificationRequest *receivedRequest;
@property (nonatomic, strong) UNMutableNotificationContent *bestAttemptContent;

@end

@implementation NotificationService

- (void)didReceiveNotificationRequest:(UNNotificationRequest *)request withContentHandler:(void (^)(UNNotificationContent * _Nonnull))contentHandler {
    self.receivedRequest = request;
    self.contentHandler = contentHandler;
    self.bestAttemptContent = [request.content mutableCopy];

    NSDictionary *userInfo = request.content.userInfo;

    // Debug: Log the entire userInfo to see what we receive
    NSLog(@"[ShopUnite NSE] Received notification with userInfo: %@", userInfo);

    // Find imageUrl in the payload
    NSString *imageUrlString = [self extractImageUrlFromUserInfo:userInfo];

    if (!imageUrlString || imageUrlString.length == 0) {
        NSLog(@"[ShopUnite NSE] No imageUrl found, delivering as-is");
        self.contentHandler(self.bestAttemptContent);
        return;
    }

    NSLog(@"[ShopUnite NSE] Found imageUrl: %@", imageUrlString);

    // Download and attach the image
    [self downloadImageAndAttach:imageUrlString];
}

- (NSString *)extractImageUrlFromUserInfo:(NSDictionary *)userInfo {
    // Expo sends data in various formats, check all possibilities
    // Log all keys for debugging
    NSLog(@"[ShopUnite NSE] userInfo keys: %@", userInfo.allKeys);

    // 1. Direct in userInfo (custom field)
    if (userInfo[@"imageUrl"] && [userInfo[@"imageUrl"] isKindOfClass:[NSString class]]) {
        NSLog(@"[ShopUnite NSE] Found imageUrl directly in userInfo");
        return userInfo[@"imageUrl"];
    }

    // 2. In "richContent" dictionary (Expo official API)
    if (userInfo[@"richContent"] && [userInfo[@"richContent"] isKindOfClass:[NSDictionary class]]) {
        NSDictionary *richContent = userInfo[@"richContent"];
        if (richContent[@"image"] && [richContent[@"image"] isKindOfClass:[NSString class]]) {
            NSLog(@"[ShopUnite NSE] Found image in richContent");
            return richContent[@"image"];
        }
    }

    // 3. In the "body" dictionary (Expo format)
    if (userInfo[@"body"] && [userInfo[@"body"] isKindOfClass:[NSDictionary class]]) {
        NSDictionary *body = userInfo[@"body"];
        if (body[@"imageUrl"]) {
            NSLog(@"[ShopUnite NSE] Found imageUrl in body");
            return body[@"imageUrl"];
        }
        if (body[@"richContent"] && [body[@"richContent"] isKindOfClass:[NSDictionary class]]) {
            NSDictionary *richContent = body[@"richContent"];
            if (richContent[@"image"]) {
                NSLog(@"[ShopUnite NSE] Found image in body.richContent");
                return richContent[@"image"];
            }
        }
    }

    // 4. In "data" dictionary
    if (userInfo[@"data"] && [userInfo[@"data"] isKindOfClass:[NSDictionary class]]) {
        NSDictionary *data = userInfo[@"data"];
        if (data[@"imageUrl"]) {
            NSLog(@"[ShopUnite NSE] Found imageUrl in data");
            return data[@"imageUrl"];
        }
    }

    // 5. In "aps" dictionary
    if (userInfo[@"aps"] && [userInfo[@"aps"] isKindOfClass:[NSDictionary class]]) {
        NSDictionary *aps = userInfo[@"aps"];
        if (aps[@"imageUrl"]) {
            NSLog(@"[ShopUnite NSE] Found imageUrl in aps");
            return aps[@"imageUrl"];
        }
    }

    // 6. In "experienceId" related Expo fields
    if (userInfo[@"experienceId"]) {
        NSLog(@"[ShopUnite NSE] This is an Expo notification (experienceId found)");
    }

    // 7. Search in all nested dictionaries for any image-related field
    for (NSString *key in userInfo.allKeys) {
        id value = userInfo[key];
        if ([value isKindOfClass:[NSDictionary class]]) {
            NSDictionary *dict = (NSDictionary *)value;
            // Check for imageUrl
            if (dict[@"imageUrl"] && [dict[@"imageUrl"] isKindOfClass:[NSString class]]) {
                NSLog(@"[ShopUnite NSE] Found imageUrl in nested dict: %@", key);
                return dict[@"imageUrl"];
            }
            // Check for image (Expo richContent style)
            if (dict[@"image"] && [dict[@"image"] isKindOfClass:[NSString class]]) {
                NSLog(@"[ShopUnite NSE] Found image in nested dict: %@", key);
                return dict[@"image"];
            }
        }
    }

    NSLog(@"[ShopUnite NSE] No image URL found in any location");
    return nil;
}

- (void)downloadImageAndAttach:(NSString *)imageUrlString {
    NSURL *imageUrl = [NSURL URLWithString:imageUrlString];

    if (!imageUrl) {
        NSLog(@"[ShopUnite NSE] Invalid URL: %@", imageUrlString);
        self.contentHandler(self.bestAttemptContent);
        return;
    }

    NSURLSessionDownloadTask *task = [[NSURLSession sharedSession] downloadTaskWithURL:imageUrl completionHandler:^(NSURL * _Nullable location, NSURLResponse * _Nullable response, NSError * _Nullable error) {

        if (error) {
            NSLog(@"[ShopUnite NSE] Download error: %@", error.localizedDescription);
            self.contentHandler(self.bestAttemptContent);
            return;
        }

        if (!location) {
            NSLog(@"[ShopUnite NSE] No download location");
            self.contentHandler(self.bestAttemptContent);
            return;
        }

        // Determine file extension from MIME type
        NSString *fileExtension = @"jpg";
        NSString *mimeType = response.MIMEType;

        if ([mimeType containsString:@"png"]) {
            fileExtension = @"png";
        } else if ([mimeType containsString:@"gif"]) {
            fileExtension = @"gif";
        } else if ([mimeType containsString:@"jpeg"] || [mimeType containsString:@"jpg"]) {
            fileExtension = @"jpg";
        }

        // Create temp file with correct extension
        NSString *tempFileName = [NSString stringWithFormat:@"%@.%@", [[NSUUID UUID] UUIDString], fileExtension];
        NSURL *tempFileUrl = [NSURL fileURLWithPath:[NSTemporaryDirectory() stringByAppendingPathComponent:tempFileName]];

        NSError *moveError = nil;
        [[NSFileManager defaultManager] moveItemAtURL:location toURL:tempFileUrl error:&moveError];

        if (moveError) {
            NSLog(@"[ShopUnite NSE] File move error: %@", moveError.localizedDescription);
            self.contentHandler(self.bestAttemptContent);
            return;
        }

        // Create notification attachment
        NSError *attachmentError = nil;
        UNNotificationAttachment *attachment = [UNNotificationAttachment attachmentWithIdentifier:@"image" URL:tempFileUrl options:nil error:&attachmentError];

        if (attachmentError) {
            NSLog(@"[ShopUnite NSE] Attachment error: %@", attachmentError.localizedDescription);
            self.contentHandler(self.bestAttemptContent);
            return;
        }

        if (attachment) {
            NSLog(@"[ShopUnite NSE] Successfully attached image!");
            self.bestAttemptContent.attachments = @[attachment];
        }

        self.contentHandler(self.bestAttemptContent);
    }];

    [task resume];
}

- (void)serviceExtensionTimeWillExpire {
    NSLog(@"[ShopUnite NSE] Time expired, delivering best attempt");
    self.contentHandler(self.bestAttemptContent);
}

@end

File: ios-extensions/NotificationService.swift
import UserNotifications

class NotificationService: UNNotificationServiceExtension {

    var contentHandler: ((UNNotificationContent) -> Void)?
    var bestAttemptContent: UNMutableNotificationContent?

    override func didReceive(_ request: UNNotificationRequest, withContentHandler contentHandler: @escaping (UNNotificationContent) -> Void) {
        self.contentHandler = contentHandler
        bestAttemptContent = (request.content.mutableCopy() as? UNMutableNotificationContent)

        guard let bestAttemptContent = bestAttemptContent else {
            contentHandler(request.content)
            return
        }

        // Buscar la URL de imagen en el payload
        // Expo envía los datos en "body" -> "data" -> "imageUrl"
        var imageUrlString: String?

        // Intentar obtener imageUrl de diferentes ubicaciones del payload
        if let data = request.content.userInfo["body"] as? [String: Any],
           let url = data["imageUrl"] as? String {
            imageUrlString = url
        } else if let url = request.content.userInfo["imageUrl"] as? String {
            imageUrlString = url
        } else if let data = request.content.userInfo["data"] as? [String: Any],
                  let url = data["imageUrl"] as? String {
            imageUrlString = url
        }

        // Si encontramos una URL de imagen, descargarla
        guard let urlString = imageUrlString,
              let imageUrl = URL(string: urlString) else {
            contentHandler(bestAttemptContent)
            return
        }

        // Descargar la imagen
        downloadImage(from: imageUrl) { attachment in
            if let attachment = attachment {
                bestAttemptContent.attachments = [attachment]
            }
            contentHandler(bestAttemptContent)
        }
    }

    override func serviceExtensionTimeWillExpire() {
        // Llamado justo antes de que la extensión sea terminada por el sistema
        if let contentHandler = contentHandler, let bestAttemptContent = bestAttemptContent {
            contentHandler(bestAttemptContent)
        }
    }

    private func downloadImage(from url: URL, completion: @escaping (UNNotificationAttachment?) -> Void) {
        let task = URLSession.shared.downloadTask(with: url) { localUrl, response, error in
            guard let localUrl = localUrl, error == nil else {
                completion(nil)
                return
            }

            // Determinar la extensión del archivo
            let fileExtension: String
            if let mimeType = response?.mimeType {
                switch mimeType {
                case "image/jpeg":
                    fileExtension = "jpg"
                case "image/png":
                    fileExtension = "png"
                case "image/gif":
                    fileExtension = "gif"
                default:
                    fileExtension = "jpg"
                }
            } else {
                fileExtension = url.pathExtension.isEmpty ? "jpg" : url.pathExtension
            }

            // Crear archivo temporal con la extensión correcta
            let tempDirectory = FileManager.default.temporaryDirectory
            let tempFile = tempDirectory.appendingPathComponent(UUID().uuidString + "." + fileExtension)

            do {
                // Mover el archivo descargado
                try FileManager.default.moveItem(at: localUrl, to: tempFile)

                // Crear el attachment
                let attachment = try UNNotificationAttachment(identifier: UUID().uuidString, url: tempFile, options: nil)
                completion(attachment)
            } catch {
                print("Error creating attachment: \(error)")
                completion(nil)
            }
        }
        task.resume()
    }
}
